crates/test-macro/src/lib.rs

```diff
@@ -34,23 +35,95 @@ pub fn wasm_bindgen_test(
     let mut leading_tokens = Vec::new();
     while let Some(token) = body.next() {
         leading_tokens.push(token.clone());
-        if let TokenTree::Ident(token) = token {
-            if token == "async" {
-                r#async = true;
+
+        match token {
+            // Parse a potential `should_panic` attribute
+            // Start by parsing the `#`
+            TokenTree::Punct(op) if op.as_char() == '#' => {
+                // We need to peek into the next token, save it before we continue.
+                let token = if let Some(token) = body.next() {
+                    leading_tokens.push(token.clone());
+                    token
+                } else {
+                    continue;
+                };
+
+                // Parse `[...]`
+                let group = match token {
+                    TokenTree::Group(group) if group.delimiter() == Delimiter::Bracket => group,
+                    _ => continue,
+                };
+
+                let mut stream = group.stream().into_iter();
+
+                // Parse `should_panic`
+                match stream.next() {
+                    Some(TokenTree::Ident(token)) if token == "should_panic" => (),
+                    _ => continue,
+                }
+
+                should_panic = Some(None);
+
+                // We are interested in the `expected` attribute or string if there is any
+                match stream.next() {
+                    // Parse the `(...)` in `#[should_panic(...)]`
+                    Some(TokenTree::Group(group))
+                        if group.delimiter() == Delimiter::Parenthesis =>
+                    {
+                        let mut stream = group.stream().into_iter();
+
+                        // Parse `expected`
+                        match stream.next() {
+                            Some(TokenTree::Ident(token)) if token == "expected" => (),
+                            _ => continue,
+                        }
+
+                        // Parse `=`
+                        match stream.next() {
+                            Some(TokenTree::Punct(op)) if op.as_char() == '=' => (),
+                            _ => continue,
+                        }
+                    }
+                    // Parse `=`
+                    Some(TokenTree::Punct(op)) if op.as_char() == '=' => (),
+                    _ => continue,
+                }
```

Comment:
I specifically designed it like this to avoid breaking `wasm-bindgen-test` on future additions to `should_panic` in Rust or custom attributes.

Also throwing compile errors here is problematic to a degree, we should actually return with an error but still print the user input even if it's wrong, because otherwise other things on the users side will break because suddenly the whole function is gone.

Happy to redesign it though if you still think it's the right way to go about it.

Reply:
I decided to throw a compile error instead of `continue`.
See https://github.com/rustwasm/wasm-bindgen/pull/3293#discussion_r1102646172.

Reply:
Should I also throw an error if we detect multiple `should_panic` attributes?
Just checked with Rust, only throws a warning there but no error (on stable there is currently no way to warn with a proc-macro (that I know of)).

Reply:
Yes, you should - the warning Rust gives says it'll become a hard error in future, so we should make it a hard error too.

Reply:
Done.